cpp/src/arrow/util/future.h

```diff
@@ -273,11 +349,134 @@ class Future {
   }
 
   /// \brief Make a finished Future<> with the provided Status.
-  template <typename E = ValueType, typename = detail::Empty::EnableIfSame<E>>
+  template <typename E = ValueType, typename = typename std::enable_if<
+                                        std::is_same<E, detail::Empty>::value>::type>
   static Future<> MakeFinished(Status s = Status::OK()) {
     return MakeFinished(E::ToResult(std::move(s)));
   }
 
+  /// \brief Consumer API: Register a callback to run when this future completes
+  ///
+  /// The callback should receive the result of the future (const Result<T>&)
+  /// For a void or statusy future this should be
+  /// (const Result<Future<detail::Empty>::ValueType>& result)
```

Comment:
... which is? Ideally, the user gets something "simple", such as `Status` or `const Status&`.

Reply:
Yes.  They receive `const Result<T>&` or `const Result<detail::Empty>&`

Reply:
I think it's preferable to keep things as generic as possible, so the empty result is preferable to the plain Status